driver_test.go

```diff
@@ -2938,3 +2939,57 @@ func TestValuerWithValueReceiverGivenNilValue(t *testing.T) {
 		// This test will panic on the INSERT if ConvertValue() does not check for typed nil before calling Value()
 	})
 }
+
+// TestContextCancelQueryWhileScan checks for race conditions that arise when
+// a query context is canceled while a user is calling rows.Scan(). The code
+// is based on database/sql TestIssue18429.
+// See https://github.com/golang/go/issues/23519
+func TestContextCancelQueryWhileScan(t *testing.T) {
+	const blob = "0123456789abcdef"
+	const sqlQuery = `SELECT id, value FROM test WHERE SLEEP(?) = 0`
+	const contextRaceIterations = 1000
+	const milliWait = 30
+	const blobSize = 64 * 1024
+	const insertRows = 64
+
+	largeBlob := strings.Repeat(blob, blobSize/len(blob))
+
+	runTests(t, dsn, func(dbt *DBTest) {
+		dbt.mustExec("CREATE TABLE test (id int, value MEDIUMBLOB) CHARACTER SET utf8")
+		for i := 0; i < insertRows; i++ {
+			dbt.mustExec("INSERT INTO test VALUES (?, ?)", i+1, largeBlob)
+		}
+
+		sem := make(chan bool, 20)
+		var wg sync.WaitGroup
+		for i := 0; i < contextRaceIterations; i++ {
+			sem <- true
+			wg.Add(1)
+			go func() {
+				defer func() {
+					<-sem
+					wg.Done()
+				}()
+				qwait := float64(time.Duration(rand.Intn(milliWait))*time.Millisecond) / float64(time.Second)
+
+				ctx, cancel := context.WithTimeout(context.Background(), time.Duration(rand.Intn(milliWait))*time.Millisecond)
+				defer cancel()
+
+				rows, _ := dbt.db.QueryContext(ctx, sqlQuery, qwait)
+				if rows != nil {
+					var b int
+					var n string
+					for rows.Next() {
+						if rows.Scan(&b, &n) == nil {
```

Comment:
This does not illustrate the core of the issue.  No need to insert sleep in query.

```go
for rows.Next() {
    if rows.Scan(&b, &n) == nil {
        s1 := string(n)
        time.Sleep(time.Millisecond*5)
        s2 := string(n)
        if s1 != s2 {
            t.Fatal("race happens")
        }
    }
}
```

Reply:
@methane: I'm afraid this is not correct. Manually cancelling the context sequentially like you're doing with your example will never trigger the race, because the race needs to happen _inside_ `rows.Scan`, so it is not possible to `cancel()` after the `Scan` to corrupt data.

That is the reason why we're adding a `SLEEP` to the SQL query: to increase the chances that the asynchronous timeout from `context.WithTimeout` will trigger the race: we need a slow query.  I have however noticed that we can work around the sleep by adding large rows to the MySQL database, which make the query slower _and_ increase the chances of the race triggering because it takes longer to `Scan` the rows. I've updated the test accordingly. :ok_hand:

Reply:
Uh, I missed `var n` must be `[]byte`, not string.

> Manually cancelling the context sequentially like you're doing with your example will never trigger the race, because the race needs to happen _inside_ `rows.Scan`, so it is not possible to `cancel()` after the `Scan` to corrupt data.

*After* cancelling context, *another* goroutine calls `driver.Rows.Close()`.  It breaks data inside `[]byte`.  Reading data inside `[]byte` is enough to trigger the race.

Reply:
Here's the stack traces of the race again:

```
WARNING: DATA RACE
Write at 0x00c0005a6000 by goroutine 105:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:197 +0x0
  github.com/go-sql-driver/mysql.(*buffer).fill()
      /home/vmg/src/gopath/src/github.com/go-sql-driver/mysql/buffer.go:46 +0x661
  github.com/go-sql-driver/mysql.(*buffer).readNext()
      /home/vmg/src/gopath/src/github.com/go-sql-driver/mysql/buffer.go:97 +0x1c5
  github.com/go-sql-driver/mysql.(*mysqlConn).readPacket()
      /home/vmg/src/gopath/src/github.com/go-sql-driver/mysql/packets.go:67 +0x25d
  github.com/go-sql-driver/mysql.(*mysqlConn).readUntilEOF()
      /home/vmg/src/gopath/src/github.com/go-sql-driver/mysql/packets.go:795 +0x45
  github.com/go-sql-driver/mysql.(*mysqlRows).Close()
      /home/vmg/src/gopath/src/github.com/go-sql-driver/mysql/rows.go:116 +0x17f
  github.com/go-sql-driver/mysql.(*textRows).Close()
      <autogenerated>:1 +0x43
  database/sql.(*Rows).close.func1()
      /usr/local/go/src/database/sql/sql.go:2989 +0x70
  database/sql.withLock()
      /usr/local/go/src/database/sql/sql.go:3097 +0x74
  database/sql.(*Rows).close()
      /usr/local/go/src/database/sql/sql.go:2988 +0x19b
  database/sql.(*Rows).awaitDone()
      /usr/local/go/src/database/sql/sql.go:2646 +0x12d

Previous read at 0x00c0005a6001 by goroutine 128:
  runtime.slicebytetostring()
      /usr/local/go/src/runtime/string.go:75 +0x0
  database/sql.asString()
      /usr/local/go/src/database/sql/convert.go:484 +0x81b
  database/sql.convertAssignRows()
      /usr/local/go/src/database/sql/convert.go:423 +0xdde
  database/sql.(*Rows).Scan()
      /usr/local/go/src/database/sql/sql.go:2955 +0x505
  github.com/go-sql-driver/mysql.TestContextCancelQueryWhileScan.func1.1()
      /home/vmg/src/gopath/src/github.com/go-sql-driver/mysql/driver_test.go:2981 +0x2a3
```

As you can see, the race happens between `(*Rows).Scan` and `(*Rows).close`. For the race to trigger, `close` needs to be called **from a background goroutine**, which is `sql.(*Rows).awaitDone()`, created by the `sql` package because we're using `context.WithTimeout`. If you `cancel()` manually, the cancellation happens concurrently and the data race will not manifest.

Reply:
> Uh, I missed var n must be []byte, not string.

This does not make the test work. The `[]byte` has already been cloned by `Scan`: https://golang.org/src/database/sql/convert.go#L259, so casting to `string` twice will never make it different. To catch the race, we must be writing to the buffer when `cloneBytes` happens, and the only way is if this happens on a background goroutine.

Reply:
Also note that we could pass a `sql.RawBytes` instead of `[]byte`, which would keep the original memory, but that would **not** trigger the issue:

```
// RawBytes is a byte slice that holds a reference to memory owned by
// the database itself. After a Scan into a RawBytes, the slice is only
// valid until the next call to Next, Scan, or Close.
type RawBytes []byte
```

If we `cancel()`, the `RawBytes` will become invalidated. This is **not** the race we're trying to fix here, that is the expected behavior from the `sql` package.

Reply:
I'm sorry, not `[]byte`, but `sql.RawBytes`.

Reply:
:sweat: Could you please read my comments more carefully? I know English can be a barrier, but we're arguing in circles here.

The contract for `RawBytes` clearly states that "the slice is only valid until the next call to Next, Scan, or Close". We know that canceling the context manually closes the `Rows`, so reading from `RawBytes` does not need to be consistent. This behavior has always been consistent in the `sql` package, so I don't think it's a good idea to be testing for that. The test I'm proposing shows a reproduction for the race that affects `[]bytes` and `string`, which is the issue that users will be running into.

Reply:
> The contract for `RawBytes` clearly states that "the slice is only valid until the next call to Next, Scan, or Close". We know that canceling the context manually closes the `Rows`, so reading from `RawBytes` does not need to be consistent.

Context may be cancelled in anytime, by anyone.  User can't prohibit context is cancelled while reading RawBytes.

So, RawBytes must be safe even context is cancelled, until user code (not `database/sql`) calls `sql.Rows.Next()` or `sql.Rows.Close()` (not `driver.Rows.Next()` or `driver.Rows.Close()).

Reply:
If you can not understand me, change this code:

```go
cancel()  // database/sql may call driver.Rows.Close() from other goroutine.
time.Sleep(time.Millisecond*5)
```

with:

```go
go func() { cancel() }() // some other goroutine or timeout may cancel the context.
time.Sleep(time.Millisecond*5)
```

Reply:
@methane: I've changed the test altogether with an explicit call to `cancel()`. I still think using `RawBytes` is overreaching into the driver contract, but hey, whatever is needed to land the actual fix for the issue. :smile: